### DISCUSSION TITLE
Migration of fabric-ca-client package to be included in fabric-gateway

### DISCUSSION THREAD

**Discussion by juannavasNexplore** [hyperledger/fabric-gateway#637] [Q&A]
As a hyperledger developer, I am migrating to the HLF v2.5 version, then I see the SDKs for v2.2 are deprecated, specifically the fabric-ca-client states in its npm registry page (https://www.npmjs.com/package/fabric-ca-client):
" Note: This API is deprecated as of Fabric v2.5. When developing applications for Hyperledger Fabric v2.4 and later, you should use the Fabric Gateway client API."

However, the fabric-gateway client api currently does not include the fabric-ca-client functionality. 

Our application is making use of the fabric-ca-client package, so I would need it to be ported to the fabric-gateway package, if fabric-ca-client is deprecated and eventually removed.

I would appreciate if you can share any rough time plan, or provide alternatives. 

Thanks!

**Comment by bestbeforetoday** [comment]
The Fabric Gateway client API (similar to the [fabric-network](https://www.npmjs.com/package/fabric-network) package that preceded it) focuses only on enabling client business applications to invoke (evaluate and submit) transactions, and receive chaincode and block events. It intentionally omits capability for administrative operations, such as chaincode deployment, channel creation, or identity registration / enrolment.

For administrative tasks, there are a few potential options:

- Use the [Fabric CLI commands](https://hyperledger-fabric-ca.readthedocs.io/en/latest/users-guide.html#fabric-ca-client).
- If deploying to Kubernetes, a Kubernetes operator designed for managing Fabric might provide the capability you need. Two alternatives that I know of are [Hyperledger Fabric Operator](https://hyperledger.github.io/bevel-operator-fabric/) and [fabric-operator](https://github.com/hyperledger-labs/fabric-operator).
- Continue using the existing [fabric-ca-client](https://www.npmjs.com/package/fabric-ca-client). While it is no longer actively developed, the Fabric CA is pretty stable in its capability so this client API should continue to work.
- Use the REST API of the Fabric CA directly with your own client. I don't know of any specific documentation of the API, but the [fabric-ca-client implementation](https://github.com/hyperledger/fabric-sdk-node/tree/main/fabric-ca-client/lib) just interacts with the Fabric CA using HTTP requests. You could inspect the implementation (or perhaps the [unit tests](https://github.com/hyperledger/fabric-sdk-node/blob/main/fabric-ca-client/test/FabricCAClient.js) to identify the correct HTTP request structure. The [Swagger definition](https://github.com/hyperledger/fabric-ca/blob/main/swagger/swagger-fabric-ca.json) of the CA REST services and [this StackOverflow post](https://stackoverflow.com/questions/78191095/accessing-fabric-ca-server-with-rest-api-authentication-error) might also be helpful.

For chaincode deployment and channel management, there is a pretty complete Golang implementation in the [fabric-admin-sdk](https://github.com/hyperledger/fabric-admin-sdk) repository. That repository also contains a Node admin implementation (based on code developed [here](https://github.com/davidkhala/fabric-common/tree/master/nodejs/admin)) with code to interact with the CA. I am not sure how stable and/or functional the Node implementation is.

**Reply by juannavasNexplore** [reply]
Ok, thanks a lot for the clarification!!

**Reply by davidkhala** [reply]
Hi @juannavasNexplore I am the author of the original npm package `khala-admin-sdk`and nice to meet you here. Recently I am working on make it more like a standard library including some tasks (below can import somehow off-topic content)
- migrate to hyperledger fabric-admin-sdk by npm package `fabric-format`, `fabrid-admin` (and possibly ca-client in discussion)
- rewrite them into ts
- rewrite end-2-end tests: originally, that is done in my repos github.com/davidkhala/delphi-fabric. To align with go parts of admin-sdk, it should depends on fabric-samples/test-network
